### PR TITLE
BigQuery: Add routines support 

### PR DESCRIPTION
**Pull request by shollyman**
This adds java support for bigquery routines, which enables API access to a persistent functions and (soon) stored procedures.  Routines exist inside of BigQuery datasets, just like tables and BQML models.

UDF documentation:
https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions
Routines REST documentation:
https://cloud.google.com/bigquery/docs/reference/rest/v2/routines

This PR adds:
CRUD operations for manipulating Routine resources.  It models the same structure used for Tables and Models.  There's a RoutineId class for the qualified reference, a RoutineInfo for metadata and a Routine class for binding to the service operations like create/update/delete.  Submessages like arguments (RoutineArgument) are also implemented as classes.

StandardSQLDataType,StandardSQLField,StandardSQLStructType veneers for communicating type information using standard sql typing.  These are used for communicating datatype information for inputs/outputs in routines.

Augments query stats:  In addition to using the API directly, you can also manipulate routines using the query language (e.g. CREATE FUNCTION ... AS ...).  Statistics provided as part of the query job include the statement type, and when using appropriate DDL queries the DDL target routine.

There's some other cleanup around typos/docs, as well as filling in missing things like statements in the StatementType enum.

### CODE REVIEW

**Comment by shollyman** [comment]
Failure is due to unrelated flaky test in BQ, sent https://github.com/googleapis/google-cloud-java/pull/5592 to address.

File: google-cloud-clients/google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/it/ITBigQueryTest.java [hunk 1]
```diff
@@ -18,13 +18,7 @@
...
 import static com.google.cloud.bigquery.JobStatus.State.DONE;
 import static com.google.common.truth.Truth.assertThat;
-import static org.junit.Assert.assertArrayEquals;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.*;
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/it/ITBigQueryTest.java]: Please revert the change to `*`

File: google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java [hunk 2]
```diff
@@ -267,6 +300,27 @@ public static ModelListOption pageToken(String pageToken) {
...
+  class RoutineListOption extends Option {
+
+    private static final long serialVersionUID = 8660294969063312498L;
+
+    private RoutineListOption(BigQueryRpc.Option option, Object value) {
+      super(option, value);
+    }
+
+    /** Returns an option to specify the maximum number of models returned per page. */
+    public static RoutineListOption pageSize(long pageSize) {
```
[the remark this answers is not in the document]
Comment (shollyman) [reply google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java]: It's probably reasonable, we extend Option but don't appear to extend each of the various typed options beyond that.  Filed https://github.com/googleapis/google-cloud-java/issues/5728 to followup on visibility more generally.

File: google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java [hunk 3]
```diff
@@ -804,6 +885,29 @@ public int hashCode() {
...
+   * String datasetId = "my_dataset_id";
+   * String routineId = "my_routine_id";
+   * RoutineId routineId = RoutineId.of(projectId, datasetId, routineId);
+   * boolean deleted = bigquery.delete(routineId);
+   * if (deleted) {
+   *   // the routine was deleted
+   * } else {
+   *   // the routine was not found
+   * }
+   * }</pre>
```
[the remark this answers is not in the document]
Comment (shollyman) [reply google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java]: done.

File: google-cloud-clients/google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/it/ITBigQueryTest.java (hunk 1 above)
Comment (shollyman) [reply google-cloud-clients/google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/it/ITBigQueryTest.java]: done.  IDE trying to be helpful and failing.
